### ISSUE TITLE
when uses more than 2 graphic card to run, it brings problem.

### ISSUE THREAD

**Issue by moliqian** [BestSonny/SSTD#19]
@BestSonny ,Hi.   when i use more than 2 graphic card to run , it wrong.
like following:
Check failed: error == cudaSuccess (77 vs. 0) an illegal memory access was encountered.
Check failed: error == status == CUBLAS_STATUS_SUCCESS (11 vs. 0)CUBLAS_STATUS_MAPPING_ERROR

I guess that, maybe "Annotated_mask_data_layer" is something wrong.  but i am not sure.
Have met the problem?  please give me some advice. Thanks in advance.

**Comment by BestSonny** [comment]
@moliqian  I can not figure out your problem from the description.

However, you may try a smaller image to see whether it solves this problem.

**Comment by moliqian** [comment]
@BestSonny hey.  When you use Annotated_data_mask_layer to train network, had you met the same problem as mine?

**Comment by BestSonny** [comment]
@moliqian No, I can train it successfully.

**Comment by moliqian** [comment]
hi, BestSonny. Thanks for your patience.
When i use one graphic card to train net, it's OK. But, more than two graphic card,  it's wrong.
Could you please tell me how many graphic card did you used?

**Comment by moliqian** [comment]
Sorry to bother you again.
I re-descript my issue. I want to do segment and detect  simultaneously  IN VOC12 ，using Caffe.  My enviroment is Ubuntu 16.04 LTS , Nvidia-driver is 375.66 , 8 gpus( 1080Ti) .  CUDA Version 8.0.61 .  cudnn5.1.10  .  I  add  the new layer(annotated_data_mask_layer) into SSD code [https://github.com/weiliu89/caffe/tree/ssd] . And using the tool "convert_annoset" to convert segmentation-detection dataset into lmdb. And then I complie caffe-ssd source code successfully. When I train net using 1 gpu, it training successfully and smoothly. When i use a big batch, it end up with a bug. When I train using more than 1 gpu , it end up with a same bug( LIke following).  

### I0121 22:13:48.079303 37268 net.cpp:283] Network initialization done.
I0121 22:13:48.079836 37268 solver.cpp:75] Solver scaffolding done.
I0121 22:13:48.082417 37268 caffe.cpp:155] Finetuning from models/VGGNet/VGG_ILSVRC_16_layers_fc_reduced.caffemodel
I0121 22:13:48.137434 37268 upgrade_proto.cpp:67] Attempting to upgrade input file specified using deprecated input fields: models/VGGNet/VGG_ILSVRC_16_layers_fc_reduced.caffemodel
I0121 22:13:48.137457 37268 upgrade_proto.cpp:70] Successfully upgraded file specified using deprecated input fields.
W0121 22:13:48.137462 37268 upgrade_proto.cpp:72] Note that future Caffe releases will only support input layers and not input fields.
I0121 22:13:48.155977 37268 net.cpp:761] Ignoring source layer drop6
I0121 22:13:48.156981 37268 net.cpp:761] Ignoring source layer drop7
I0121 22:13:48.156987 37268 net.cpp:761] Ignoring source layer fc8
I0121 22:13:48.156991 37268 net.cpp:761] Ignoring source layer prob
I0121 22:13:48.211648 37268 upgrade_proto.cpp:67] Attempting to upgrade input file specified using deprecated input fields: models/VGGNet/VGG_ILSVRC_16_layers_fc_reduced.caffemodel
I0121 22:13:48.211671 37268 upgrade_proto.cpp:70] Successfully upgraded file specified using deprecated input fields.
W0121 22:13:48.211674 37268 upgrade_proto.cpp:72] Note that future Caffe releases will only support input layers and not input fields.
I0121 22:13:48.229640 37268 net.cpp:761] Ignoring source layer drop6
I0121 22:13:48.230568 37268 net.cpp:761] Ignoring source layer drop7
I0121 22:13:48.230574 37268 net.cpp:761] Ignoring source layer fc8
I0121 22:13:48.230578 37268 net.cpp:761] Ignoring source layer prob
I0121 22:13:48.269497 37268 parallel.cpp:392] GPUs pairs 0:1, 2:3, 0:2
I0121 22:13:48.973143 37268 annotated_data_mask_layer.cpp:67] output data size: 2,3,320,320
I0121 22:13:48.974512 37268 annotated_data_mask_layer.cpp:86] output mask size: 2,1,320,320
I0121 22:13:50.220502 37268 annotated_data_mask_layer.cpp:67] output data size: 2,3,320,320
I0121 22:13:50.221127 37268 annotated_data_mask_layer.cpp:86] output mask size: 2,1,320,320
I0121 22:13:51.416030 37268 annotated_data_mask_layer.cpp:67] output data size: 2,3,320,320
I0121 22:13:51.417455 37268 annotated_data_mask_layer.cpp:86] output mask size: 2,1,320,320
I0121 22:13:52.096508 37268 parallel.cpp:425] Starting Optimization
I0121 22:13:52.096639 37268 solver.cpp:294] Solving VGG_VOC0712_SSD_320x320_train
I0121 22:13:52.096762 37268 solver.cpp:295] Learning Rate Policy: multistep
F0121 22:13:52.532143 37371 math_functions.cu:121] Check failed: status == CUBLAS_STATUS_SUCCESS (11 vs. 0)  CUBLAS_STATUS_MAPPING_ERROR
*** Check failure stack trace: ***
F0121 22:13:52.532294 37372 syncedmem.hpp:31] Check failed: error == cudaSuccess (77 vs. 0)  an illegal memory access was encounteredF0121 22:13:52.532418 37373 syncedmem.hpp:18] Check failed: error == cudaSuccess (77 vs. 0)  an illegal memory access was encountered
*** Check failure stack trace: ***
F0121 22:13:52.532294 37372 syncedmem.hpp:31] Check failed: error == cudaSuccess (77 vs. 0)  an illegal memory access was encounteredF0121 22:13:52.532418 37373 syncedmem.hpp:18] Check failed: error == cudaSuccess (77 vs. 0)  an illegal memory access was encountered
*** Check failure stack trace: ***
F0121 22:13:52.532294 37372 syncedmem.hpp:31] Check failed: error == cudaSuccess (77 vs. 0)  an illegal memory access was encounteredF0121 22:13:52.532418 37373 syncedmem.hpp:18] Check failed: error == cudaSuccess (77 vs. 0)  an illegal memory access was encounteredF0121 22:13:52.533569 37268 syncedmem.hpp:18] Check failed: error == cudaSuccess (77 vs. 0)  an illegal memory access was encountered
*** Check failure stack trace: ***
    @     0x7f4d876045cd  google::LogMessage::Fail()
    @     0x7f4d876045cd  google::LogMessage::Fail()
    @     0x7f4d876045cd  google::LogMessage::Fail()
    @     0x7f4d876045cd  google::LogMessage::Fail()
    @     0x7f4d87606433  google::LogMessage::SendToLog()
    @     0x7f4d87606433  google::LogMessage::SendToLog()
    @     0x7f4d87606433  google::LogMessage::SendToLog()
    @     0x7f4d87606433  google::LogMessage::SendToLog()
    @     0x7f4d8760415b  google::LogMessage::Flush()
    @     0x7f4d8760415b  google::LogMessage::Flush()
    @     0x7f4d87606e1e  google::LogMessageFatal::~LogMessageFatal()
    @     0x7f4d87606e1e  google::LogMessageFatal::~LogMessageFatal()
    @     0x7f4d8760415b  google::LogMessage::Flush()
    @     0x7f4d8760415b  google::LogMessage::Flush()
    @     0x7f4d87e63bb1  caffe::SyncedMemory::mutable_cpu_data()
    @     0x7f4d87e63bb1  caffe::SyncedMemory::mutable_cpu_data()
    @     0x7f4d87606e1e  google::LogMessageFatal::~LogMessageFatal()
    @     0x7f4d87606e1e  google::LogMessageFatal::~LogMessageFatal()
    @     0x7f4d87e70671  caffe::Blob<>::Reshape()
    @     0x7f4d87e70671  caffe::Blob<>::Reshape()
    @     0x7f4d87f7065a  caffe::caffe_gpu_asum<>()
    @     0x7f4d87e630b0  caffe::SyncedMemory::~SyncedMemory()
    @     0x7f4d87e70b1a  caffe::Blob<>::Reshape()
    @     0x7f4d87f6b2ce  caffe::ComputeConfLossGPU<>()
    @     0x7f4d87d19452  boost::detail::sp_counted_impl_p<>::dispose()
    @     0x7f4d87f6439d  caffe::SoftmaxWithLossLayer<>::Backward_gpu()
    @     0x7f4d87e70bcc  caffe::Blob<>::Blob()
    @     0x7f4d87ec1eff  caffe::MineHardExamples<>()
    @     0x7f4d87e6464a  boost::detail::sp_counted_base::release()
    @     0x7f4d87cd364b  caffe::Net<>::BackwardFromTo()
    @     0x7f4d87e70907  caffe::Blob<>::Reshape()
    @     0x7f4d87cd36af  caffe::Net<>::Backward()
    @     0x7f4d87f6b02f  caffe::ComputeConfLossGPU<>()
    @     0x7f4d87df5642  caffe::MultiBoxLossLayer<>::Forward_cpu()
    @     0x7f4d87e46873  caffe::SoftmaxLayer<>::Reshape()
    @     0x7f4d87e9bc8c  caffe::Solver<>::Step()
    @     0x7f4d87d35191  caffe::SoftmaxWithLossLayer<>::Reshape()
    @     0x7f4d87ec1eff  caffe::MineHardExamples<>()
    @     0x7f4d87cd2d67  caffe::Net<>::ForwardFromTo()
    @     0x7f4d87ca6fc6  caffe::P2PSync<>::InternalThreadEntry()
    @     0x7f4d87df591d  caffe::MultiBoxLossLayer<>::Forward_cpu()
    @     0x7f4d87e77105  caffe::InternalThread::entry()
    @     0x7f4d87df5642  caffe::MultiBoxLossLayer<>::Forward_cpu()
    @     0x7f4d87cd30d7  caffe::Net<>::Forward()
    @     0x7f4d7d2255d5  (unknown)
    @     0x7f4d87cd2d67  caffe::Net<>::ForwardFromTo()
    @     0x7f4d7ac616ba  start_thread
    @     0x7f4d87cd30d7  caffe::Net<>::Forward()
    @     0x7f4d87cd2d67  caffe::Net<>::ForwardFromTo()
    @     0x7f4d87e9bc80  caffe::Solver<>::Step()
    @     0x7f4d85e8141d  clone
    @     0x7f4d87e9bc80  caffe::Solver<>::Step()
    @              (nil)  (unknown)

I think there may be 3 possible where i did wrong. 
1. I create a wrong lmdb dataset. But when i print net.label , net.mask , net.data ,   they are all correct. 
     I still can run correct in a small batch size using one gpu.
2. There is something wrong in annotated_data_mask_layer. But I still can run correct in a small batch size using one gpu.
3. Maybe there is something wrong in cuda and cudnn .   Maybe there is a misaligned address issue in cuda or cudnn, lead to "Check failed: status == CUBLAS_STATUS_SUCCESS (11 vs. 0)  CUBLAS_STATUS_MAPPING_ERROR" and "Check failed: error == cudaSuccess (77 vs. 0)  an illegal memory access was encountered".

How can i fix the bug "Check failed: error == cudaSuccess (77 vs. 0)  an illegal memory access was enc  an illegal memory access was encountered".
Could you please tell me how many graphic card did you used? which version cuda and cudnn did you use? Thanks in advance.

**Comment by BestSonny** [comment]
@moliqian  Make sure to use our release caffe.

**Comment by moliqian** [comment]
Hi, @BestSonny I download SSTD , when i "make runtest" ,  it comes with an issue. Like following,

### [----------] 1 test from LayerFactoryTest/3, where TypeParam = caffe::GPUDevice<double>
[ RUN      ] LayerFactoryTest/3.TestCreateLayer
F0123 17:03:44.816540 12895 db_leveldb.cpp:16] Check failed: status.ok() Failed to open leveldb 
IO error: /LOCK: Permission denied
*** Check failure stack trace: ***
    @     0x2b8e46df6daa  (unknown)
    @     0x2b8e46df6ce4  (unknown)
    @     0x2b8e46df66e6  (unknown)
    @     0x2b8e46df9687  (unknown)
    @     0x2b8e492b5d27  caffe::db::LevelDB::Open()
    @     0x2b8e49353cd3  caffe::DataReader<>::Body::InternalThreadEntry()
    @     0x2b8e4934e185  caffe::InternalThread::entry()
    @     0x2b8e489c8a4a  (unknown)
    @     0x2b8e4a348184  start_thread
    @     0x2b8e4a65bffd  (unknown)
    @              (nil)  (unknown)
make: *** [runtest] Aborted (core dumped)

Whether there is something I am missing ?

**Comment by moliqian** [comment]
Hi, @BestSonny .  I download SSTD , make and make test successfully. when i only test annotated_data_mask_layer (using the method [(https://www.cnblogs.com/louyihang-loves-baiyan/p/5260863.html]) , it doesn't pass with seraval error. Like following, I doubt that this SSTD release is something wrong. Could you please download SSTD, and "make", "make test" , "make runtest" , to see whether this SSTD release is some wrong ?

The difference between std::min(b*0.1 + 0.2, 1.0) and label_data[cur_bbox*8+p] is 1.2, which exceeds this->eps_, where
std::min(b*0.1 + 0.2, 1.0) evaluates to 0.20000000000000001,
label_data[cur_bbox*8+p] evaluates to -1, and
this->eps_ evaluates to 9.9999999747524271e-07.
src/caffe/test/test_annotated_data_mask_layer.cpp:380: Failure
The difference between std::min(b*0.1 + 0.2, 1.0) and label_data[cur_bbox*8+p] is 1.2, which exceeds this->eps_, where
std::min(b*0.1 + 0.2, 1.0) evaluates to 0.20000000000000001,
label_data[cur_bbox*8+p] evaluates to -1, and
this->eps_ evaluates to 9.9999999747524271e-07.
src/caffe/test/test_annotated_data_mask_layer.cpp:382: Failure
Value of: label_data[cur_bbox*8+7]
  Actual: -1
Expected: a % 2
Which is: 0
[  FAILED  ] AnnotatedDataMaskLayerTest/0.TestReshapeLMDB, where TypeParam = caffe::CPUDevice<float> (217 ms)
[ RUN      ] AnnotatedDataMaskLayerTest/0.TestReadCropTrainLevelDB
[       OK ] AnnotatedDataMaskLayerTest/0.TestReadCropTrainLevelDB (245 ms)
[ RUN      ] AnnotatedDataMaskLayerTest/0.TestReadCropTestLMDB
[       OK ] AnnotatedDataMaskLayerTest/0.TestReadCropTestLMDB (56 ms)
[ RUN      ] AnnotatedDataMaskLayerTest/0.TestReadCropTrainLMDB
[       OK ] AnnotatedDataMaskLayerTest/0.TestReadCropTrainLMDB (49 ms)
[ RUN      ] AnnotatedDataMaskLayerTest/0.TestReadCropTrainSequenceSeededLMDB
[       OK ] AnnotatedDataMaskLayerTest/0.TestReadCropTrainSequenceSeededLMDB (43 ms)
[ RUN      ] AnnotatedDataMaskLayerTest/0.TestReadCropTrainSequenceUnseededLMDB
[       OK ] AnnotatedDataMaskLayerTest/0.TestReadCropTrainSequenceUnseededLMDB (42 ms)
[ RUN      ] AnnotatedDataMaskLayerTest/0.TestReadCropTrainSequenceSeededLevelDB
[       OK ] AnnotatedDataMaskLayerTest/0.TestReadCropTrainSequenceSeededLevelDB (432 ms)
[----------] 12 tests from AnnotatedDataMaskLayerTest/0 (6439 ms total)

[----------] Global test environment tear-down
[==========] 48 tests from 4 test cases ran. (23949 ms total)
[  PASSED  ] 32 tests.
[  FAILED  ] 16 tests, listed below:
[  FAILED  ] AnnotatedDataMaskLayerTest/0.TestReadLevelDB, where TypeParam = caffe::CPUDevice<float>
[  FAILED  ] AnnotatedDataMaskLayerTest/0.TestReshapeLevelDB, where TypeParam = caffe::CPUDevice<float>
[  FAILED  ] AnnotatedDataMaskLayerTest/0.TestReadLMDB, where TypeParam = caffe::CPUDevice<float>
[  FAILED  ] AnnotatedDataMaskLayerTest/0.TestReshapeLMDB, where TypeParam = caffe::CPUDevice<float>
[  FAILED  ] AnnotatedDataMaskLayerTest/1.TestReshapeLevelDB, where TypeParam = caffe::CPUDevice<double>
[  FAILED  ] AnnotatedDataMaskLayerTest/1.TestReadLevelDB, where TypeParam = caffe::CPUDevice<double>
[  FAILED  ] AnnotatedDataMaskLayerTest/1.TestReshapeLMDB, where TypeParam = caffe::CPUDevice<double>
[  FAILED  ] AnnotatedDataMaskLayerTest/1.TestReadLMDB, where TypeParam = caffe::CPUDevice<double>
[  FAILED  ] AnnotatedDataMaskLayerTest/2.TestReshapeLevelDB, where TypeParam = caffe::GPUDevice<float>
[  FAILED  ] AnnotatedDataMaskLayerTest/2.TestReadLevelDB, where TypeParam = caffe::GPUDevice<float>
[  FAILED  ] AnnotatedDataMaskLayerTest/2.TestReadLMDB, where TypeParam = caffe::GPUDevice<float>
[  FAILED  ] AnnotatedDataMaskLayerTest/2.TestReshapeLMDB, where TypeParam = caffe::GPUDevice<float>
[  FAILED  ] AnnotatedDataMaskLayerTest/3.TestReshapeLMDB, where TypeParam = caffe::GPUDevice<double>
[  FAILED  ] AnnotatedDataMaskLayerTest/3.TestReshapeLevelDB, where TypeParam = caffe::GPUDevice<double>
[  FAILED  ] AnnotatedDataMaskLayerTest/3.TestReadLevelDB, where TypeParam = caffe::GPUDevice<double>
[  FAILED  ] AnnotatedDataMaskLayerTest/3.TestReadLMDB, where TypeParam = caffe::GPUDevice<double>

16 FAILED TESTS
make: *** [runtest_annotated_data_mask] Error 1

**Comment by BestSonny** [comment]
@moliqian You can skip 'run test' and directly train the network.

**Comment by moliqian** [comment]
I have solved this problem. In the data_tramsform.cpp, 
the line 490 : **EncodeCVMatToDatum(crop_img, "jpg", crop_datum);**
change into:
    **if(cv_img.channels() == 3) {
      EncodeCVMatToDatum(crop_img, "jpg", crop_datum);
    } else {
      EncodeCVMatToDatum(crop_img, "png", crop_datum);
    }**

@BestSonny thanks for your helping!
